### PR TITLE
WP_Theme_JSON: sanitize: Optimize schema calculation by processing only used blocks

### PR DESCRIPTION
**Pull request by mreishus**
Trac ticket: https://core.trac.wordpress.org/ticket/62187

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

### CODE REVIEW

File: src/wp-includes/class-wp-theme-json.php [hunk 1]
```diff
@@ -989,6 +989,9 @@ protected static function sanitize( $input, $valid_block_names, $valid_element_n
 		$schema_styles_blocks   = array();
 		$schema_settings_blocks = array();
 
+		$used_blocks = self::get_used_blocks( $input );
+		$valid_used_blocks = array_intersect( $used_blocks, $valid_block_names );
```
Comment (mukeshpanchal27) [inline src/wp-includes/class-wp-theme-json.php]:
```suggestion
		$used_blocks       = self::get_used_blocks( $input );
		$valid_used_blocks = array_intersect( $used_blocks, $valid_block_names );
```

**Comment by mukeshpanchal27** [comment]
I did XHG profiling and in my test it shows the below result in which it shows some function call increased instead of improvement 

Screenshot 2024-10-29 at 2 40 28 PM

**Comment by mreishus** [comment]
Thank you, confirmed. This patch makes the out of the box experience worse. The improvement seems to come in after changing to the twentytwentyfour theme and installing plugins like WooCommerce and Jetpack. That is, increasing the number of blocks rendered and blocks registered. Will have to think about it.

### Condition: 10.29.nightly.2025.theme

| Save Name   | Timestamp           | p50 (ms) | Number of Requests |
|-------------|----------------------|----------|---------------------|
| trunk       | 2024-10-29 16:53:36 | 37.12    | 250                |
| with-patch  | 2024-10-29 16:55:33 | 41.01    | 250                |

### Condition: 10.29.nightly.2024.theme

| Save Name   | Timestamp           | p50 (ms) | Number of Requests |
|-------------|----------------------|----------|---------------------|
| trunk       | 2024-10-29 16:57:15 | 50.41    | 250                |
| patch       | 2024-10-29 16:57:41 | 49.80    | 250                |

### Condition: 10.29.nightly.2024.theme.with.woocommerce

| Save Name   | Timestamp           | p50 (ms) | Number of Requests |
|-------------|----------------------|----------|---------------------|
| trunk       | 2024-10-29 17:01:06 | 115.02   | 250                |
| patch       | 2024-10-29 17:01:53 | 112.53   | 250                |

### Condition: 10.29.nightly.2024.theme.with.woocommerce.and.jetpack

| Save Name   | Timestamp           | p50 (ms) | Number of Requests |
|-------------|----------------------|----------|---------------------|
| trunk       | 2024-10-29 17:16:57 | 133.73   | 250                |
| patch       | 2024-10-29 17:17:45 | 123.37   | 250                |

**Comment by mreishus** [comment]
Closing this PR after reviewing performance data. While the patch improves performance for complex installations with many registered blocks (e.g. WooCommerce + Jetpack), it introduces overhead in the default WordPress configuration. Since core needs to optimize primarily for the common case, this approach isn't the right solution. Will explore alternative optimizations that don't negatively impact the default installation.

**Pull request closed without merging**